### ISSUE TITLE
Add AppImage deploy tools to build system

### ISSUE THREAD

**Issue by dalboris** [vgc/vgc#58]
The idea is to make it easy to create AppImages. Something like:

```
mkdir build && cd $_
cmake \
  -DLINUXDEPLOYQT_PATH=<path-to-linuxdeployqt> \
  ..
make appimages
```

This would create one AppImage for each VGC application. If we want to deploy for one app only:

```
make appimage vgcillustration
```

**Comment by probonopd** [comment]
Although not written in CMake, this should give you an idea:
https://github.com/vgc/vgc/compare/master...probonopd:patch-1.diff

**Comment by dalboris** [comment]
Thanks, your patch is indeed very useful as a reference. Is it possible to pass the desired app version to linuxdeployqt without setting the `VERSION` environment variable? Not a big deal, but it seems that it would be nicer if this was a documented optional argument to pass to linuxdeployqt.

**Comment by probonopd** [comment]
No, but you can do something like `( VERSION=... ./linuxdeployqt*.AppImage ...` if you don't want to "pollute" your environment.

**Comment by dalboris** [comment]
Alright, the AppImage was successfully generated by TravisCI and uploaded to the server, but when trying to run it I got the following:

```
$ ./vgcillustration-alpha-2019-10-28.0-x86_64.AppImage 
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: initfsencoding: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007fc05b0fb780 (most recent call first):
Aborted (core dumped)
```

This is obviously related to Python not being correctly embedded, which was somewhat expected since I didn't yet copy the python standard libraries to the AppImage.

**Comment by probonopd** [comment]
Yes, Python and its modules and libraries need to be copied into the AppDir, too.

**Comment by dalboris** [comment]
Yes, I merely hoped that just starting the interpreter (not importing any modules except `sys`), might at least work with only libpython[...].so.

Do you think automatically copying python modules and libs for apps that depend on Python might be someday be on the scope of linuxdeployqt or appimagetool?

**Comment by dalboris** [comment]
Some useful references below.

Relevant pages from the official Python documentation. Most importantly, there are configuration functions in the C API which can be called before `Py_initialize()` to control where the python libraries live without having to use environment variables such as `PYTHONPATH` or `PYTHONHOME`:
  - https://docs.python.org/3/library/site.html
  - https://docs.python.org/3/library/sys.html#sys.base_prefix
  - https://docs.python.org/3/library/sys.html#sys.path
  - https://docs.python.org/3/extending/embedding.html
  - https://docs.python.org/3/c-api/init.html

> **void Py_SetProgramName(const wchar_t \*name)**
> This function should be called before Py_Initialize() is called for the first time, if it is called at all. It tells the interpreter the value of the argv[0] argument to the main() function of the program (converted to wide characters). This is used by Py_GetPath() and some other functions below to find the Python run-time libraries relative to the interpreter executable. The default value is 'python'. The argument should point to a zero-terminated wide character string in static storage whose contents will not change for the duration of the program’s execution. No code in the Python interpreter will change the contents of this storage.
> 
> **wchar\* Py_GetProgramName()**
> Return the program name set with Py_SetProgramName(), or the default. The returned string points into static storage; the caller should not modify its value.
> 
> **wchar_t\* Py_GetPrefix()**
> Return the prefix for installed platform-independent files. This is derived through a number of complicated rules from the program name set with Py_SetProgramName() and some environment variables; for example, if the program name is '/usr/local/bin/python', the prefix is '/usr/local'. The returned string points into static storage; the caller should not modify its value. This corresponds to the prefix variable in the top-level Makefile and the --prefix argument to the configure script at build time. The value is available to Python code as sys.prefix. It is only useful on Unix. See also the next function.
> 
> **wchar_t\* Py_GetExecPrefix()**
> Return the exec-prefix for installed platform-dependent files. This is derived through a number of complicated rules from the program name set with Py_SetProgramName() and some environment variables; for example, if the program name is '/usr/local/bin/python', the exec-prefix is '/usr/local'. The returned string points into static storage; the caller should not modify its value. This corresponds to the exec_prefix variable in the top-level Makefile and the --exec-prefix argument to the configure script at build time. The value is available to Python code as sys.exec_prefix. It is only useful on Unix.
> 
> Background: The exec-prefix differs from the prefix when platform dependent files (such as executables and shared libraries) are installed in a different directory tree. In a typical installation, platform dependent files may be installed in the /usr/local/plat subtree while platform independent may be installed in /usr/local.
> 
> Generally speaking, a platform is a combination of hardware and software families, e.g. Sparc machines running the Solaris 2.x operating system are considered the same platform, but Intel machines running Solaris 2.x are another platform, and Intel machines running Linux are yet another platform. Different major revisions of the same operating system generally also form different platforms. Non-Unix operating systems are a different story; the installation strategies on those systems are so different that the prefix and exec-prefix are meaningless, and set to the empty string. Note that compiled Python bytecode files are platform independent (but not independent from the Python version by which they were compiled!).
> 
> System administrators will know how to configure the mount or automount programs to share /usr/local between platforms while having /usr/local/plat be a different filesystem for each platform.
> 
> **wchar_t\* Py_GetProgramFullPath()**
> Return the full program name of the Python executable; this is computed as a side-effect of deriving the default module search path from the program name (set by Py_SetProgramName() above). The returned string points into static storage; the caller should not modify its value. The value is available to Python code as sys.executable.
> 
> **wchar_t\* Py_GetPath()**
> Return the default module search path; this is computed from the program name (set by Py_SetProgramName() above) and some environment variables. The returned string consists of a series of directory names separated by a platform dependent delimiter character. The delimiter character is ':' on Unix and Mac OS X, ';' on Windows. The returned string points into static storage; the caller should not modify its value. The list sys.path is initialized with this value on interpreter startup; it can be (and usually is) modified later to change the search path for loading modules.
> 
> **void Py_SetPath(const wchar_t \*)**
> Set the default module search path. If this function is called before Py_Initialize(), then Py_GetPath() won’t attempt to compute a default search path but uses the one provided instead. This is useful if Python is embedded by an application that has full knowledge of the location of all modules. The path components should be separated by the platform dependent delimiter character, which is ':' on Unix and Mac OS X, ';' on Windows.
> 
> This also causes sys.executable to be set to the program full path (see Py_GetProgramFullPath()) and for sys.prefix and sys.exec_prefix to be empty. It is up to the caller to modify these if required after calling Py_Initialize().
> 
> The path argument is copied internally, so the caller may free it after the call completes.
>
> **void Py_SetPythonHome(const wchar_t \*home)**
> Set the default “home” directory, that is, the location of the standard Python libraries. See PYTHONHOME for the meaning of the argument string.
> 
> The argument should point to a zero-terminated character string in static storage whose contents will not change for the duration of the program’s execution. No code in the Python interpreter will change the contents of this storage.
> 
> **w_char\* Py_GetPythonHome()**
> Return the default “home”, that is, the value set by a previous call to Py_SetPythonHome(), or the value of the PYTHONHOME environment variable if it is set.

Looking at what FreeCAD does may be useful. They also embed Python, use Qt, and deploy as an AppImage on Linux:
  - https://www.freecadweb.org/wiki/Appimage
  - https://github.com/FreeCAD/FreeCAD/releases
  - https://github.com/FreeCAD/FreeCAD/blob/1995f9d0bac63820c5c42ac0075c91a49cbad119/src/Base/Interpreter.cpp#L512

```
const char* InterpreterSingleton::init(int argc,char *argv[])
{
    if (!Py_IsInitialized()) {
#if PY_MAJOR_VERSION >= 3
#if PY_MINOR_VERSION >= 5
        Py_SetProgramName(Py_DecodeLocale(argv[0],NULL));
#else
        Py_SetProgramName(_Py_char2wchar(argv[0],NULL));
#endif
#else
        Py_SetProgramName(argv[0]);
#endif
        // There is a serious bug in VS from 2010 until 2013 where the file descriptor for stdin, stdout or stderr
        // returns a valid value for GUI applications (i.e. subsystem = Windows) where it shouldn't.
        // This causes Python to fail during initialization.
        // A workaround is to use freopen on stdin, stdout and stderr. See the class Redirection inside main()
        // https://bugs.python.org/issue17797#msg197474
        //
        Py_Initialize();
        const char* virtualenv = getenv("VIRTUAL_ENV");
        if (virtualenv) {
            PyRun_SimpleString(
                "# Check for virtualenv, and activate if present.\n"
                "# See https://virtualenv.pypa.io/en/latest/userguide/#using-virtualenv-without-bin-python\n"
                "import os\n"
                "import sys\n"
                "base_path = os.getenv(\"VIRTUAL_ENV\")\n"
                "if not base_path is None:\n"
                "    activate_this = os.path.join(base_path, \"bin\", \"activate_this.py\")\n"
                "    exec(open(activate_this).read(), {'__file__':activate_this})\n"
            );
        }
        PyEval_InitThreads();
#if PY_MAJOR_VERSION >= 3
        size_t size = argc;
        wchar_t **_argv = new wchar_t*[size];
        for (int i = 0; i < argc; i++) {
#if PY_MINOR_VERSION >= 5
            _argv[i] = Py_DecodeLocale(argv[i],NULL);
#else
            _argv[i] = _Py_char2wchar(argv[i],NULL);
#endif
        }
        PySys_SetArgv(argc, _argv);
#else
        PySys_SetArgv(argc, argv);
#endif
        PythonStdOutput::init_type();
        this->_global = PyEval_SaveThread();
    }
#if PY_MAJOR_VERSION >= 3
    PyGILStateLocker lock;
#if PY_MINOR_VERSION >= 5
    return Py_EncodeLocale(Py_GetPath(),NULL);
#else
    return _Py_wchar2char(Py_GetPath(),NULL);
#endif
#else
    return Py_GetPath();
#endif
}
```

Finally, I found it insightful to look into binary files and search for hardcoded paths. Okteta is a nice tool for doing this: `sudo apt install okteta`

**Comment by probonopd** [comment]
What's wrong with setting `PYTHONPATH` or `PYTHONHOME`?

**Comment by dalboris** [comment]
End users are not supposed to do that. End users are just supposed to run the `vgcillustration` binary, which knows where the python libs live, since they are shipped together with it. Basically, you can see python as a "resource" used by `vgcillustration`.

**Comment by probonopd** [comment]
Yes, I meant to export those variables in the `AppRun` inside the `AppDir`, so the user doesn't need to do anything.

https://github.com/AppImage/AppImageKit/blob/801e789390d0e6848aef4a5802cd52da7f4abafb/src/AppRun.c#L169

**Comment by dalboris** [comment]
Indeed, in the specific case of an AppImage, this would work too I suppose (though is it easy to easy to modify the AppRun when using `linuxdeployqt -appimage`?).

But in the general case, the binary should know by itself where to find the python libraries. Like when installing different versions of Python, if we call, say, `/usr/bin/python` or `~/miniconda/bin/python`, each of those know where to find its corresponding runtime libraries, regardless of the environment variables. Similarly, `vgcillustration` is conceptually a fancy python interpreter, and should know where to find its Python runtime libraries, regardless of environment variables.

**Comment by probonopd** [comment]
> But in the general case, the binary should know by itself where to find the python libraries. 

Actually, this would be great behavior for Python itself. Have you thought about upstreaming it?

**Comment by dalboris** [comment]
I've been looking at this today, and the main issue seems to be that there is no reliable cross-platform way for shared libraries to know their own location. On the other hand, libraries can detect the location of the main executable calling them, via some combination of `/proc/...` and `argv[0]`, see for example the [implementation](https://github.com/qt/qtbase/blob/c92cb78b6d7c6590c79ec0e2df4dd577617bd3e1/src/corelib/kernel/qcoreapplication.cpp#L2312) of [`QCoreApplication::applicationFilePath()`](https://doc.qt.io/qt-5/qcoreapplication.html#applicationFilePath).

This is annoying because in the general case (not so much in the case of an AppImage where we can have full control of all paths), the library may be in a completely different location as the application using the library. More to the point, a library has in general no control or knowledge of where the applications using the libraries are or will be located.

In order to alleviate this problem, libraries who rely on runtime resources, such as Qt or Python, seem to resort to a mix of the following mechanisms:

- Compile hard-coded paths into the library, so it knows where it lives and where its runtime resources live. The obvious problem of this technique is that it fails if we relocate the library, unless we binary-patch the library after relocation. Qt does that for sure (see [QLibraryInfo](https://doc.qt.io/qt-5/qlibraryinfo.html)), and I don't think Python does that, see:
```
$ cd ~/Qt/5.12.5/gcc_64/lib
$ grep "/home/boris" libQt5Core.so.5.12.5
Binary file libQt5Core.so.5.12.5 matches

$ cd ~/miniconda3/lib
$ grep "/home/boris" libpython3.6m.so.1.0 
(no output)
```

- Try to guess/search the resource paths at runtime using complex heuristics based on typical conventions where libraries are typically installed (similar to CMake's find_package()). This is obviously error-prone: if several installations of the same library exist, it could mistakenly detect the resources of the other installation. Python does that at a last resort. The risk is somehow mitigated in the case of Python since the name of the folder containing python libraries folder have the version in it (e.g., `lib/python3.6`), but still it could find an installation of Python 3.6.2 instead of 3.6.1, or another installation with the exact same version but different compile flags, for example. So it's definitely not a clean solution, but somehow "magically works" in typical cases.

-  Require (or allow) **applications** using the library to specify (or override) the library's resource paths via a function call to the library. In other words, since the library itself has no reliable way to find its own resources, it asks some help from the application using the library. I don't think Qt supports this. VGC currently does this via [`vgc::core::setBasePath()`](https://github.com/vgc/vgc/blob/95152e9425155e901cffa6e3d37dcb246ec5fc09/libs/vgc/core/paths.h#L26). Python supports this via the following functions, ordered from "explicit" to "perform some magic":
    - [`Py_SetPath()`](https://docs.python.org/3/c-api/init.html#c.Py_SetPath): explicitely sets sys.path, that is, the path(s) where runtime modules are located
    - [`Py_SetPythonHome()`](https://docs.python.org/3/c-api/init.html#c.Py_SetPythonHome): sets the Python prefix, from which Python will infer sys.path by concatenating the prefix with paths like `lib/pythonX.Y`, `lib/pythonXY.zip`, `lib/pythonX.Y/site-packages`, etc.
    - [`Py_SetProgramName()`](https://docs.python.org/3/c-api/init.html#c.Py_SetProgramName): sets the name/path of the current executable, from which Python will infer the PythonHome. This works for the typical cases where the application is the python interpreter itself (<prefix>/bin/python) or other python tools (<prefix>/bin/pip) living alongside the python libraries (<prefix>/lib/pythonX.Y). But it doesn't work in the more complex cases where an application embeds python and live in a completely different location, for example if you compile VGC locally (say in `~/test/build-vgc/`) but use a system-installed python.

-  Require (or allow) **applications or users** to specify (or override) the paths by placing a configuration file alongside the application binary. For example, , Qt uses [qt.conf](https://doc.qt.io/qt-5/qt-conf.html), and on Windows Python uses a [._pth file](https://docs.python.org/3/using/windows.html#finding-modules).

-  Require (or allow) **shell wrappers or users** to specify (or override) the paths by specifying environment variables. I don't think Qt supports that. Python uses [PYTHONHOME](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHOME) and [PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH), which work like their C API counterparts. This is what we could use in the AppRun, but using the C API is in my opinion slightly cleaner, and more importantly, works in more use cases.

In the case of VGC, I have now started to implement a vgc.conf technique, similar to qt.conf. I was confused about this Qt mechanism the first time I encountered it, but with hindsight I find it the cleanest approach. It gives full control to applications, makes it easy for either packagers or users to edit the paths as required, makes it easy to fine-tune the location of resources based on platform conventions (e.g., `VGC Illustration.app/Contents/Resources/vgc` on macOS, and `/usr/share/vgc` on Linux), without hard-coding these conventions in the library itself, and the paths in the vgc.conf file can be made either relative (for bundles which include their dependent libraries), or absolute (for applications using system libraries). The library itself is then fully relocatable, and avoids obscure heuristics. It just needs some minimal collaboration from applications using the library. Optionally, applications which don't want to use a vgc.conf file can still explicitly call the C API `vgc::core::setBasePath()`, using their own methods/heuristics to determine the correct basePath.

With respect to your question, I think the Python folks are then perfectly aware of the situation, and decided to go with the current approach knowing full well the alternatives. The situation is in fact even more complicated with the possibility to create Python virtual environments, a situation for which I haven't yet fully looked at the mechanisms in place.

**Comment by dalboris** [comment]
Note: The qt.conf / vgc.conf approach still always requires some C-level collaboration as well: the main program must pass `argv[0]` to the library, which in case of Qt, is done via the constructor of the `QCoreApplication`. This is also somewhat similar to the Python C API requiring `Py_SetProgramName(argv[0])` for some of the magic to work.

**Comment by probonopd** [comment]
Isn't it as simple as finding out where the `python` executable lives, and calculate everything else relative to that location?

**Comment by dalboris** [comment]
No it's not as simple, that's the point:
1. lib/libpython.so has no way to easily find where bin/python lives, unless bin/python tells lib/libpython where bin/python lives.
2. bin/vgcillustration doesn't even use bin/python at all! In fact, I don't even have to deploy bin/python in the AppImage. When you embed Python, you don't call bin/python, the interpreter program. You directly make C calls to lib/libpython.so, the library, and *become* the interpreter. And you can live very far from libpython.so.

**Comment by probonopd** [comment]
> lib/libpython.so has no way to easily find where bin/python lives, unless bin/python tells lib/libpython where bin/python lives.

That's what I meant when I was referring to "upstreaming". The upstream `python` could do this by default.

> `bin/vgcillustration` doesn't even use bin/python at all!

True. It would have to tell `lib/libpython.so` where it lives.

**Comment by dalboris** [comment]
`bin/python` does already tell `lib/libpython.so` where `bin/python` lives, as far as I understand, by telling it the value of `argv[0]` :) That's why you don't need to set any environment variables when calling, say, `/usr/bin/python` vs `~/miniconda3/bin/python`. They all find their respective python runtime libraries (`lib/pythonX.Y/...`) correctly.

My issue was indeed not when calling `bin/python`, but when calling `bin/vgcillustration`, which may live in a completely different location. Telling `lib/libpython.so` where `bin/vgcillustration` lives doesn't help `lib/libpython.so` know where `lib/pythonX.Y/` lives. The latter information has to be given by `bin/vgcillustration` via `Py_SetPythonHome()` or `Py_SetPath()`, or given by the environment via `PYTHONHOME` or `PYTHONPATH`, as I detailed above. Anyway, I now have a clear plan on how to handle this: call `Py_SetPythonHome(path)`, with `path` given in an ASCII configuration file living alongside `bin/vgcillustration`, or alongside any other application linking to `lib/libvgccore.so`.

**Comment by probonopd** [comment]
> My issue was indeed not when calling bin/python, but when calling bin/vgcillustration, which may live in a completely different location. 

Wouldn't it be an option to assume that they live in the same location?

**Comment by dalboris** [comment]
No, and in fact, they often don't. When you clone `github.com/vgc/vgc` and call `mkdir build && cd build && cmake .. && make`, then `vgcillustration` lives in `~/vgc/build/Release/bin`, while `python` lives wherever CMake found it.

**Comment by dalboris** [comment]
@probonopd Is there a way to tell linuxdeployqt to analyze and fix the r(un)paths of binaries inside the AppDir besides the app-binary and its dependents?

In the TravisCI log [here](https://travis-ci.org/vgc/vgc/jobs/607216745), you can see that I am printing the dependencies of all binaries in the AppDir, before and after runnning linuxdeployqt.

Some binaries are correctly handled by linuxdeployqt, for example:

```
./usr/bin/vgcillustration:
  NEEDED               libvgcwidgets.so
  NEEDED               libvgccore.so
  NEEDED               libQt5Widgets.so.5
  NEEDED               libQt5Gui.so.5
  NEEDED               libQt5Core.so.5
  NEEDED               libstdc++.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  RPATH                /home/travis/build/vgc/vgc/build/Release/lib:/opt/python/3.7.1/lib:/opt/qt512/lib
```

becomes:

```
./usr/bin/vgcillustration:
  NEEDED               libvgcwidgets.so
  NEEDED               libvgccore.so
  NEEDED               libQt5Widgets.so.5
  NEEDED               libQt5Gui.so.5
  NEEDED               libQt5Core.so.5
  NEEDED               libstdc++.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  RUNPATH              $ORIGIN/../lib
```

and:

```
./usr/lib/libvgccore.so:
  NEEDED               libpython3.7m.so.1.0
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  RPATH                /opt/python/3.7.1/lib
```

becomes:

```
./usr/lib/libvgccore.so:
  NEEDED               libpython3.7m.so.1.0
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  RUNPATH              $ORIGIN
```

However, some *.so files are not reachable from the app-binary when following the chain of dependencies, because they are not referenced in any ELF files. Instead, they are loaded at runtime by the Python interpreter, if/when we interpret strings such "import vgc.core". These *.so files don't seem to be touched by linuxdeployqt. For example:

```
./usr/lib/python3.7/vgc/core.cpython-37m-x86_64-linux-gnu.so:
  NEEDED               libvgccore.so
  NEEDED               libpython3.7m.so.1.0
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  RPATH                /home/travis/build/vgc/vgc/build/Release/lib:/opt/python/3.7.1/lib
```

is untouched by linuxdeployqt, while I would like it to become:

```
./usr/lib/python3.7/vgc/core.cpython-37m-x86_64-linux-gnu.so:
  NEEDED               libvgccore.so
  NEEDED               libpython3.7m.so.1.0
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  RUNPATH              $ORIGIN/../..
```

Of course, I could fix up those myself, but it would seem like a lot of duplicated logic from what is already available in linuxdeployqt. It would be nice if linuxdeployqt would fix all *.so files in the appdir, regardless of whether they are reachable from the chain of dependency from the app-binary. Alternatively, it would be nice to let us specify additional binary files to fix.

Thanks!

(Note: there is also another shared library, ./usr/lib/libvgcscene.so, which is not touched by linuxdeployqt either, because it it actually not used anymore by the app. I am planning to delete it this module later, so this one is no big deal.)

**Comment by dalboris** [comment]
Note 2: the AppImage can be downloaded [at this link](https://www.vgc.io/releases/pr/238/2019-11-04.1/vgcillustration-alpha-gh58b-2019-11-04.1-x86_64.AppImage)

Note 3: the AppImage seems to work correctly despite these incorrect paths, most likely because the required libraries are either already loaded, or because the loader searched in the runpaths specified by already loaded binaries.

Example:

![image](https://user-images.githubusercontent.com/4809739/68141868-a565c700-ff2e-11e9-9c36-2cefd1562f79.png)

**Comment by dalboris** [comment]
Note 4: if you wish to test the embedded Python console: Enter = newline; Ctrl+Enter = interpret the code block. All output is currently printed to the bash console outside the app.

**Comment by probonopd** [comment]
> However, some *.so files are not reachable from the app-binary when following the chain of dependencies, because they are not referenced in any ELF files. Instead, they are loaded at runtime by the Python interpreter

`linuxdeployqt` has no chance of detecting those. You need to manually supply them to `linuxdeployqt` like this:

https://gitlab.com/inkscape/inkscape/blob/7d29e37a2aedad6b658d0f641a4e92cd0a237414/packaging/appimage/generate.sh#L71-142

Be sure to somehow identify which ones you need to bundle.

**Comment by dalboris** [comment]
Perfect, thanks! This `-executable` seems what I need, I thought it was just for the main binaries, not other `*.so` files. However, any chance for an option of linuxdeployqt which is basically a shorthand for "add all `*.so`  files in the appdir as an executable"?

**Comment by probonopd** [comment]
> However, any chance for an option of linuxdeployqt which is basically a shorthand for "add all *.so files in the appdir as an executable"?

A pull request for this would certainly be appreciated.

**Comment by dalboris** [comment]
@probonopd Ok, I'll see if I can try to implement this in the next following weeks (I'm now sidetracked on other priorities). Would `-all-executables` be a good name? This would be equivalent to adding all ELF files in the appdir as an argument to `-executable` (not just *.so files, but also actual executables). I see that linuxdeployqt already has some logic to find all these files, because they are printed as "used to determine architecture", so it shouldn't be too complicated to implement.

**Comment by probonopd** [comment]
That would be great @dalboris. Maybe we should even do this by default, without its own flag?

**Comment by dalboris** [comment]
@probonopd Yes, I agree, this seems like the better default. Perhaps a good design is by default to target all executables, unless `-executable` is specified, in which case only those executables explicitly specified are targeted. I don't see any obvious backward-incompatibility this would cause, any thoughts yourself?

**Comment by probonopd** [comment]
Yes, I think this would be a good approach.
